hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/OzoneClientProducer.java

```diff
@@ -83,65 +75,20 @@ public OzoneClient createClient() throws WebApplicationException,
     return client;
   }
 
-  @PreDestroy
-  public void destroy() throws IOException {
-    client.close();
-  }
-
   private OzoneClient getClient(OzoneConfiguration config)
       throws WebApplicationException {
     OzoneClient ozoneClient = null;
     try {
-      SignatureInfo signatureInfo = signatureProcessor.parseSignature();
-
-      String stringToSign = "";
-      if (signatureInfo.getVersion() == Version.V4) {
-        stringToSign =
-            StringToSignProducer.createSignatureBase(signatureInfo, context);
-      }
-
-      String awsAccessId = signatureInfo.getAwsAccessId();
-      validateAccessId(awsAccessId);
-
-      UserGroupInformation remoteUser =
-          UserGroupInformation.createRemoteUser(awsAccessId);
-      if (OzoneSecurityUtil.isSecurityEnabled(config)) {
-        LOG.debug("Creating s3 auth info for client.");
-
-        if (signatureInfo.getVersion() == Version.NONE) {
-          throw MALFORMED_HEADER;
-        }
-
-        OzoneTokenIdentifier identifier = new OzoneTokenIdentifier();
-        identifier.setTokenType(S3AUTHINFO);
-        identifier.setStrToSign(stringToSign);
-        identifier.setSignature(signatureInfo.getSignature());
-        identifier.setAwsAccessId(awsAccessId);
-        identifier.setOwner(new Text(awsAccessId));
-        if (LOG.isTraceEnabled()) {
-          LOG.trace("Adding token for service:{}", omService);
-        }
-        Token<OzoneTokenIdentifier> token = new Token(identifier.getBytes(),
-            identifier.getSignature().getBytes(StandardCharsets.UTF_8),
-            identifier.getKind(),
-            omService);
-        remoteUser.addToken(token);
 
-      }
+      this.remoteUser = UserGroupInformation.getCurrentUser();
       ozoneClient =
-          remoteUser.doAs((PrivilegedExceptionAction<OzoneClient>) () -> {
-            return createOzoneClient();
-          });
-    } catch (OS3Exception ex) {
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("Error during Client Creation: ", ex);
-      }
-      throw wrapOS3Exception(ex);
-    } catch (Exception e) {
+          OzoneClientCache.getOzoneClientInstance(omServiceID,
```

Comment:
Thanks @avijayanhwx for reviewing this and for your comments.  Removing unnecessary unused `omService` `@injected` variable and `validateAccessId` method (both were moved to UgiFilter.java - thanks!).  Similarly removing `createOzoneClient` method, perhaps was left in because `@VisibleForTesting`  annotation however not used in s3 unit tests.

Reply:
I have one question I see we are completely removing the old code. So, once after this feature comes in there is no way for us to use the rpc based code. (As OzoneClientCache always returns single client)

Reply:
The s3 gateway Grpc client only supports the `GrpcOmTransport`.  The previous `Hadoop3OmTransport` is unavailable as the `OmTransportFactory` selects the OmTransport through the ServiceProvider (set to `GrpcOmTransport` in the META-INF file for the s3gateway).

Right, once the feature comes in the client (s3 gateway) does not support the hadoop rpc based code, however the server (OzoneManager) supports both Grpc and Hadoop rpc based s3 gateways.  During the instability period (issues, whatever as it matures) the server can support both.  In that case, for the client, it can be swapped with a prior release hadoop rpc support.  Let me know what you think.

Reply:
>  In that case, for the client, it can be swapped with a prior release hadoop rpc support
Not understood what is meant by this and it is not clear how can we use hadoop rpc with change in code to use single client?

Reply:
Yes, meaning of that line is not clear.  Thanks.  The change in code does NOT support hadoop RPC for the s3 gateway based on the cached client (single) and the OmTransport (service provider for factory selecting GrpcOmTransport).

>   In that case, for the client, it can be swapped with a prior release hadoop rpc support

To be clear, I meant that we could use a prior version (pre Grpc s3g) of the gateway that supported hadoop rpc in the cluster.  

The server supports both s3 gateways, hadoop rpc and Grpc.

Reply:
> The previous `Hadoop3OmTransport` is unavailable as the `OmTransportFactory` selects the OmTransport
> through the ServiceProvider (set to `GrpcOmTransport` in the META-INF file for the s3gateway).

The META-INF file can list multiple implementations.  Wouldn't it be possible to choose between those based on Ozone configuration, and hence keep both old and new transports in the same version of S3G?  I think that would make it easier for users to switch, rather than trying to install both versions of Ozone (may not be possible with some installers / package managers).

Reply:
Thanks @adoroszlai.  Selecting the transport through configuration was considered, but was not used because the design does not support another OMTransport.  I agree that choosing a transport through configuration is easier to upgrade from the hadoop rpc s3g to Grpc s3g, however the design does not support multiple OmTransports at this point - it does not support hadoop rpc.  As @bharatviswa504 brought up, the single ozone client persistent connection is the main reason for not supporting another transport.

Reply:
@neils-dev we should consider this maybe as a part of separate Jira but usually, having such flexibility also allows users to switch back to older impl in the events of bugs/vulnerabilities etc.

Reply:
@neils-dev 
If we have a config to choose which transport, and then by changing OzoneCacheClient to return cached client if it is Grpc, else a new Client like before. In this way we could support both correct? Not sure if I am missing something here.

Reply:
Thanks @bharatviswa504 and @swagle .  Since the s3gateway gRPC does not support another transport it is not configurable.  I don't know how easy it would be to try to support another transport - what else it impacts other than an interface for the s3g `OzoneClientProducer`, in terms of the client and endpoints.  As @swagle suggested we could open another jira to explore this.  Part of this other jira would see if we can select the OzoneClientProducer and OmTransport through configuration (have to see how it impacts the CDI injected variables as well).

Reply:
I am fine with doing this is new Jira. @neils-dev Can you open a Jira for this, so that this does not get missed out.

From my understanding if we have config to use grpc/hadoop rpc, using that config if we create RpcClient to return same for Grpc, new for hadoop Rpc we should be good. Not sure if there are more things to cover.

Reply:
Thanks @bharatviswa504.  I've opened sub-task HDDS-5813 for looking into added hadoop rpc based ozone client support (as original, new connection for every request).  Documented in that sub-task the configurable OmTransport and configurable ozone client, as discussed here.